drivers/mfd/rpi-rp2040-gpio-bridge.c

```diff
@@ -0,0 +1,1155 @@
+
+	ret = devm_gpiochip_add_data(dev, &priv_data->gc, priv_data);
+	if (ret) {
+		crypto_free_shash(priv_data->shash);
+		return dev_err_probe(dev, ret,
+				     "Could not add data to gpiochip\n");
+	}
+
+	priv_data->bypass_cache =
+		of_property_read_bool(client->dev.of_node, "bypass-cache");
+	dev_info(dev, "%s() bypass-cache: %s\n", __func__,
+		 priv_data->bypass_cache ? "true" : "false");
+
+	// Optionally configure fast_xfer if RP1 is being used
+	if (of_parse_phandle_with_args(client->dev.of_node, "fast_xfer-gpios",
+				       "#gpio-cells", 0, &of_args[0]) ||
+	    of_parse_phandle_with_args(client->dev.of_node, "fast_xfer-gpios",
+				       "#gpio-cells", 1, &of_args[1])) {
+		dev_info(dev, "Could not parse fast_xfer-gpios phandles\n");
+		goto cleanup;
+	}
+
+	if (of_args[0].np != of_args[1].np) {
+		dev_info(
+			dev,
+			"fast_xfer-gpios are not provided by the same controller\n");
+		goto cleanup;
+	}
+	dn = of_args[0].np;
+	if (!of_device_is_compatible(dn, "raspberrypi,rp1-gpio")) {
+		dev_info(dev, "fast_xfer-gpios controller is not an rp1\n");
+		goto cleanup;
+	}
+	if (of_args[0].args_count != 2 || of_args[1].args_count != 2) {
+		dev_info(dev, "of_args count is %d\n", of_args[0].args_count);
+		goto cleanup;
+	}
+
+	if (of_property_read_u32_index(client->dev.of_node,
+				       "fast_xfer_recv_gpio_base", 0,
+				       &priv_data->fast_xfer_recv_gpio_base)) {
+		dev_info(dev, "Could not read fast_xfer_recv_gpio_base\n");
+		goto cleanup;
+	}
+
+	priv_data->fast_xfer_gpios =
+		devm_gpiod_get_array_optional(dev, "fast_xfer", GPIOD_ASIS);
+	if (!priv_data->fast_xfer_gpios) {
+		dev_info(dev, "Could not acquire fast_xfer-gpios\n");
+		goto cleanup;
+	}
+
+	priv_data->fast_xfer_data_index = of_args[0].args[0];
+	priv_data->fast_xfer_clock_index = of_args[1].args[0];
+	priv_data->fast_xfer_requires_i2c_lock = of_property_read_bool(
+		client->dev.of_node, "fast_xfer_requires_i2c_lock");
+
+	dev_info(dev, "%s() fast_xfer data: %u clock: %u i2c: %s\n", __func__,
+		 priv_data->fast_xfer_data_index,
+		 priv_data->fast_xfer_clock_index,
+		 priv_data->fast_xfer_requires_i2c_lock ? "true" : "false");
+
+	priv_data->gpio_base = of_iomap(dn, 0);
+	if (IS_ERR_OR_NULL(priv_data->gpio_base)) {
+		dev_info(&client->dev, "%s() unable to map gpio_base\n",
+			 __func__);
+		priv_data->gpio_base = NULL;
+		devm_gpiod_put_array(dev, priv_data->fast_xfer_gpios);
+		priv_data->fast_xfer_gpios = NULL;
+		goto cleanup;
+	}
+
+	priv_data->rio_base = of_iomap(dn, 1);
+	if (IS_ERR_OR_NULL(priv_data->rio_base)) {
+		dev_info(&client->dev, "%s() unable to map rio_base\n",
+			 __func__);
+		priv_data->rio_base = NULL;
+		iounmap(priv_data->gpio_base);
+		priv_data->gpio_base = NULL;
+		devm_gpiod_put_array(dev, priv_data->fast_xfer_gpios);
+		priv_data->fast_xfer_gpios = NULL;
+		goto cleanup;
+	}
+
+cleanup:
+	if (of_args[0].np)
+		of_node_put(of_args[0].np);
+	if (of_args[1].np)
+		of_node_put(of_args[1].np);
+	return 0;
```

Comment:
So this is a success path? `cleanup` implied to me at least that it was an error path. `node_put` or similar would be more obvious.